### PR TITLE
Unit tests for Promtail

### PR DESCRIPTION
**Pull request by slim-bean**
Adding some channels to watch for proper closing of goroutines
Adding some unit tests around target.go

### CODE REVIEW

File: pkg/promtail/position.go [hunk 1]
```diff
@@ -29,14 +29,15 @@ func (cfg *PositionsConfig) RegisterFlags(flags *flag.FlagSet) {
...
+	logger       log.Logger
+	cfg          PositionsConfig
+	mtx          sync.Mutex
+	positions    map[string]int64
+	quit         chan struct{}
+	quitComplete chan struct{}
 }
 
-type positionsFile struct {
+type PositionsFile struct {
```
Comment (tomwilkie) [inline pkg/promtail/position.go]: Why did you need to expose this?  Capital letters at the beginning of a type make it package-public.

Comment (slim-bean) [reply pkg/promtail/position.go]: because i was loading the positions file in the unit test to verify the contents and debated between just duplicating the struct in the unit test or exposing it here.  I don't have a strong argument for one way vs the other and didn't see the harm in exposing it so picked that.

File: pkg/promtail/position.go [hunk 2]
```diff
@@ -115,12 +122,14 @@ func readPositionsFile(filename string) (map[string]int64, error) {
 	buf, err := ioutil.ReadFile(filepath.Clean(filename))
 	if err != nil {
 		if os.IsNotExist(err) {
-			return map[string]int64{}, nil
+			positions := map[string]int64{}
+			err = writePositionFile(filename, positions)
+			return positions, err
```
Comment (tomwilkie) [inline pkg/promtail/position.go]: Why do you want to write a empty file if it doesn't exist?

File: pkg/promtail/target.go [hunk 3]
```diff
@@ -191,6 +206,9 @@ type tailer struct {
 
 	path string
 	tail *tail.Tail
+
+	quit         chan struct{}
+	quitComplete chan struct{}
```
Comment (tomwilkie) [inline pkg/promtail/target.go]: `done`

File: pkg/promtail/position.go (hunk 2 above)
Comment (slim-bean) [reply pkg/promtail/position.go]: ahh, this was leftover from some very early debugging I was doing, saving the empty file serves no purpose, I will remove this

File: pkg/promtail/target.go [hunk 4]
```diff
@@ -252,11 +271,26 @@ func (t *tailer) run() {
...
+	}
+	level.Debug(t.logger).Log("path", t.path, "current_position", pos)
+	t.positions.Put(t.path, pos)
+	return nil
+}
+
 func (t *tailer) stop() error {
+	close(t.quit)
+	<-t.quitComplete
 	return t.tail.Stop()
```
Comment (tomwilkie) [inline pkg/promtail/target.go]: Is the extra `quit` channel necessary, or can we use `t.tail.Stop()`?

File: pkg/promtail/position.go (hunk 1 above)
Comment (tomwilkie) [reply pkg/promtail/position.go]: You should be able to use the same type in a unit test, the test should be in the same package.****

Comment (slim-bean) [reply pkg/promtail/position.go]: ahh, you are correct, I had a collision with a variable named positionsFile and misunderstood the error, I will return the struct to private

File: pkg/promtail/target.go (hunk 4 above)
Comment (slim-bean) [reply pkg/promtail/target.go]: this channels is necessary to update the current position with any changes since the last `positionSyncPeriod`, otherwise they are lost

File: pkg/promtail/target.go [hunk 5]
```diff
@@ -75,8 +76,16 @@ func NewTarget(logger log.Logger, handler EntryHandler, positions *Positions, pa
 	for _, p := range matches {
 		dirs[filepath.Dir(p)] = struct{}{}
 	}
+
+	//If no files exist yet watch the directory specified in the path
```
Comment (tomwilkie) [inline pkg/promtail/target.go]: Comments should have a space after the `//` and a full stop at the end.

File: pkg/promtail/target.go [hunk 6]
```diff
@@ -75,8 +76,16 @@ func NewTarget(logger log.Logger, handler EntryHandler, positions *Positions, pa
...
+	if matches == nil {
+		//TODO does this work if the path is a directory?
```
Comment (tomwilkie) [inline pkg/promtail/target.go]: No TODOs in merged code please.

File: pkg/promtail/target.go [hunk 7]
```diff
@@ -212,31 +231,32 @@ func newTailer(logger log.Logger, handler EntryHandler, positions *Positions, pa
...
-
 	level.Info(t.logger).Log("msg", "start tailing file", "filename", t.path)
 	positionSyncPeriod := t.positions.cfg.SyncPeriod
 	positionWait := time.NewTicker(positionSyncPeriod)
-	defer positionWait.Stop()
+
+	defer func() {
+		level.Info(t.logger).Log("msg", "stopping tailing file", "filename", t.path)
+		positionWait.Stop()
+		_ = t.markPosition()
```
Comment (tomwilkie) [inline pkg/promtail/target.go]: please log the error.

File: pkg/promtail/target_test.go [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+}
+
+type TestClient struct {
+	log      log.Logger
+	messages []string //FIXME should this be an array of pointers to strings?
+}
+
+func (c *TestClient) Handle(ls model.LabelSet, t time.Time, s string) error {
+	c.messages = append(c.messages, s)
+	//level.Info(c.log).Log("msg", "received log", "log", s)
```
Comment (tomwilkie) [inline pkg/promtail/target_test.go]: Don't commit commented out code.
